Selfhosting

3036662 opened the pull request:
IF LIBRUM_SELFHOSTED != true - nothing will be changed.
----------------
IF  LIBRUM_SELFHOSTED == true

1. UserBlobStorage and BooksBlobStorage classes will be substituted by UserLocalStorage and BooksLocalStorage with Dependecy Injection mechanism
2. EmailSender.cs will use variable  SMTPMailFrom to fill email sender author.
3. AIService.cs will send simple string informing that OpenAi not available if no API token exists, otherwise it will act normal.
4. UserConroller will use it's own http  methods to show simple "password reset" form.
5. Program.cs will create local database structure if not exists.
PS.
BookRepository.cs string to Double parsing method was changed due problems with running on linux (it was throwing exception)
------------------------
I would like you to take a look, I'm pretty shure we can work it out to run perfect!
For now we can run this server as linux service in fully self-hosted mode using install instruction from my fork.
At the same time azure services will continue their normal work if not self-hosted.

DavidLazarescu commented:
Thank you, this looks great. I'll look over it tomorrow and give you detailed feedback on it.

DavidLazarescu commented on src/Application/Utility/EmailSender.cs:
```diff
@@ -47,10 +54,24 @@ public async Task SendEmailConfirmationEmail(User user, string token)
+		if (_configuration["LIBRUM_SELFHOSTED"] == "true"){
+			var domain =_configuration["CleanUrl"];
+			var encodedToken=System.Web.HttpUtility.HtmlEncode(token);
+			resetLink = $"{domain}/user/resetPassword?email={user.Email}&token={encodedToken}";
```
I don't understand what you are doing here. Could you give me an example?

3036662 replied:
For example:
1. self-hosted server is listening to https  127.0.0.1:5001 , 
2. it's exposed to network as mybook.com (standart port 443)
3. _configuration["CleanUrl"]  = https://mybooks.com, 
4. user will get email with password  reset link starting with mybooks.com/user/resetPassword.....  (not   127.0.0.1:5001).  
---------------
PS. its possible that server isn't aware about  its public url . Server running on 127.0.0.1:5001 and exposed to web by nginx.
So , the _configuration["CleanUrl"] is used to build "password reset" link with "public url".

DavidLazarescu replied:
I understand. Thanks!

DavidLazarescu reviewed (changes requested):
This looks really good! I added some notes to your code, please look over them and change the PR respectively.

3036662 commented:
Hi, thanks, I'll do everything on Monday.

3036662 commented on src/Application/Managers/BookLocalStorageManager.cs:
```diff
@@ -0,0 +1,156 @@
+    public BookLocalStorageManager()
+    {
+        baseDir=System.Environment.GetFolderPath(System.Environment.SpecialFolder.Personal);
+		dataDir= baseDir+"/librum_srorage";
```
fixed

3036662 commented on src/Application/Managers/BookLocalStorageManager.cs:
```diff
@@ -0,0 +1,156 @@
+public class BookLocalStorageManager : IBookBlobStorageManager
+{
+    private readonly string _bookCoverPrefix = "cover_";
+	private string dataDir;
```
fixed

3036662 commented on src/Application/Managers/BookLocalStorageManager.cs:
```diff
@@ -0,0 +1,156 @@
+    {
+		  //if already exists
+		var filename=coversDir+"/"+guid;
+		if ( System.IO.File.Exists(filename)){
```
fixed

3036662 commented on src/Application/Managers/UserLocalStorageManager.cs:
```diff
@@ -0,0 +1,92 @@
+    public async Task ChangeProfilePicture(string guid, MultipartReader reader)
+    {
+        //if already exists
+		var filename=profilesDir+"/"+guid;
```
fixed

3036662 replied:
fixed

3036662 commented on src/Application/Managers/UserLocalStorageManager.cs:
```diff
@@ -0,0 +1,92 @@
+    public UserLocalStorageManager()
+    {
+        baseDir=System.Environment.GetFolderPath(System.Environment.SpecialFolder.Personal); // $HOME folder 
+	    dataDir= baseDir+"/librum_srorage";
```
fixed

3036662 commented on src/Application/Managers/UserLocalStorageManager.cs:
```diff
@@ -0,0 +1,92 @@
+public class UserLocalStorageManager : IUserBlobStorageManager
+{
+    private readonly string _profilePicturePrefix = "profilePicture_";
+    private string dataDir;
```
moved to local variable

3036662 commented on src/Application/Services/AIService.cs:
```diff
@@ -34,6 +34,12 @@ public class AiService : IAiService
     {
+		//  return error mesage if no OpenAIToken exists for self-hosted mode, dont throw exception
+	    if (_configuration["LIBRUM_SELFHOSTED"] == "true" &&  String.IsNullOrEmpty(_configuration["OpenAIToken"])){
+            await context.Response.WriteAsync("data: " + "OpenAI temporary unavailable" + "\n");
```
done

3036662 reviewed (commented):
Nice!

DavidLazarescu commented:
Btw. were you able to run the self-hosted version on your machine? We still need to resolve all of the configuration lookup stuff like JWTKey, ... that are requested via e.g. `configuration["JWTKey"]`

3036662 commented:
> Btw. were you able to run the self-hosted version on your machine? We still need to resolve all of the configuration lookup stuff like JWTKey, ... that are requested via e.g. `configuration["JWTKey"]`

Yes, I've tested everything , I'll give you full explanation in 30 min.

3036662 commented:
> Btw. were you able to run the self-hosted version on your machine? We still need to resolve all of the configuration lookup stuff like JWTKey, ... that are requested via e.g. `configuration["JWTKey"]`

Sorry for delay. 
So, to run normally you need:
1. install dependencies
2. build with dotnet restore & dotnet build
3. Install server
-------------------------
normally to run server on linux you should run it as service, same does mysql and other servers, but it possible to run it simply with command line, first I'll talk about "classic" way:
1. create user for server running
2. create .service file for systemd
3. configure "secret variables" like "JWTKey" etc. - service secret configuration should be store in linux /etc folder in separate file, so ```systemctl show``` command will not show them.
4. create folder in "special user" home directory and install server
5. manage file permissions, to forbid other users to read from server folder and no write permissions, of course.
6. create ssl certificate
7. when everything is done, you can run it as usual service - systemctl start server-name
----------------------------------
i will push now self-hosted- build and install  instruction draft.
-----------------------------------
The way to run it from command line - for testing:
1. build and restore
2. create ssl certificate
3. configure server ports to listen
4. simply export all secret variables 
```export secretVariable=secretValue```
5. run with dotnet run

DavidLazarescu commented:
Ok, I understand.

It would be good if the README.md of the server would contain a detailed section on this. If you could write the general outline of it, I could improve it and add examples, ...

3036662 commented:
> Ok, I understand.
> 
> It would be good if the README.md of the server would contain a detailed section on this. If you could write the general outline of it, I could improve it and add examples, ...


I've added a manpage and self-hosted-install.md  drafts to new folder "self-hosting" , i've also added config files examples.
Please, take a look, so we can make them perfect.
I've tested the algorithm few times on different configuration, but still something may be missing for some linux versions.

DavidLazarescu commented:
Looks great! I'll have more time in an hour or so, but I skimmed over it and it looks very detailed, which is great.

I really appreciate your work.

DavidLazarescu commented:
In your guide:

`apt-get install dotnet-sdk-7.0 dotnet-7.0  dotnet-host openssl mariadb-server dotnet-7.0 dotnet-aspnetcore-runtime-7.0 openssl`
can be reduced to `sudo apt install dotnet-sdk-7.0 openssl mariadb-server`

and `install librum-server.service -m660  /etc/systemd/system/` gives me the error: `install: cannot stat 'librum-server.service': No such file or directory` on ubuntu

3036662 commented:
> can be reduced to
it is not always absolutely reliable to entrust the resolution of dependencies to the package manager, but there should be no problems with this in ubuntu, but you can shorten it, he will pull up the necessary packages himself


> and `install librum-server.service -m660 /etc/systemd/system/`
``` install self-hosting/librum-server.service -m660  /etc/systemd/system/ ```
.service file is located in subfolder ```self-hosting```

DavidLazarescu commented:
> .service file is located in subfolder `self-hosting` , try this, in case of problems, tell me, i will test install ubuntu on virtual environment and try to install.

Ah, I see. It seems like `PR_techSelection` contains the correct changes but `gitHubFork_selfHosted` contains the readme. I have cloned the repository with the readme which didn't have the selfhosted folder.

Is there a reason for having these two different branches, or can they be combined into one?

3036662 commented:
> Is there a reason for having these two different branches, or can they be combined into one?
For now it looks like we don't need gitHubFork_selfHosted branche. I'll deal with it.

DavidLazarescu commented:
Also, your step to create a new user fails with the following errors:
![image](https://github.com/Librum-Reader/Librum-Server/assets/69865187/a07ceb1c-f44a-44d3-a1f0-89d58e9b7a90)

3036662 commented:
> Also, your step to create a new user fails

Looks like you need to do it another way in ubuntu, try this
useradd -r -g librum-server -d /var/lib/librum-server --shell /usr/sbin/nologin

DavidLazarescu commented:
And where exactly is `INSTALL.md` mentioned in `install -m664 INSTALL.md /var/lib/librum-server/srv`?

3036662 commented:
> I think you could change it to something like: `sudo /usr/sbin/useradd -r -g librum-server -d /var/lib/librum-server -s /usr/sbin/nologin -c "Server for librum-reader" librumserver`

I'ts just linux-destribution - specific stuff, we can change everything to fit ubuntu, for me no diffirence, we already have rpm binary packages for our distro)
For now, i dont have an installed ubuntu, but i can test it with ubuntu if needed.

3036662 commented:
> And where exactly is `INSTALL.md` mentioned in `install -m664 INSTALL.md /var/lib/librum-server/srv`?

fixed instruction, it was renamed to self-hosted-install.md

DavidLazarescu commented:
> > I think you could change it to something like: `sudo /usr/sbin/useradd -r -g librum-server -d /var/lib/librum-server -s /usr/sbin/nologin -c "Server for librum-reader" librumserver`
> 
> I'ts just linux-destribution - specific stuff, we can change everything to fit ubuntu, for me no diffirence, we already have rpm binary packages for our distro) For now, i dont have an installed ubuntu, but i can test it with ubuntu if needed.

Yes, I think we should aim the instructions for ubuntu and provide specifics for other platforms when needed. I can do that later, for now I am just trying to run the self-hosted server on my ubuntu machine.

On ubuntu, `/etc/my.cnf.d/server.cnf` does not exist, there are `conf.d/          debian.cnf       debian-start     mariadb.cnf      mariadb.conf.d/  my.cnf           my.cnf.fallback ` in the `/etc/mysql` folder

3036662 commented:
> my.cnf

Notmally mysql settings are located in my.cnf , but it can differ depending on version and distro and package version.
I'm already downloading ubuntu)

DavidLazarescu commented:
> > my.cnf
> 
> Notmally mysql settings are located in my.cnf , but it can differ depending on version and distro and package version. I'm already downloading ubuntu)

I see, I found `/etc/mysql/mariadb.conf.d/50-server.cnf` which contains the `bind-address` which you mentioned. I'll continue with this file for now then.

3036662 commented:
I,ve tested build&run process in Ubuntu 22.04, added instructions.

DavidLazarescu commented:
Ok I have followed the instructions. I had some problems with `sudo mysql_secure_installation` but I changed my password and was able to complete it.

I think you miss spelled this: `vim /etc/librum-reader/librum-reader.conf`, you probably meant `vim /etc/librum-server/librum-server.conf`

DavidLazarescu commented:
Also, don't you need to add `export LIBRUM_SELFHOSTED=true` to the run.sh script?

DavidLazarescu commented:
And how exactly is the server supposed to find the config values in `vim /etc/librum-server/librum-server.conf`? Starting the application manually fails for me since it can't find the JWT token.

3036662 commented:
> I think you miss spelled this: `vim /etc/librum-reader/librum-reader.conf`

Yes, thanks. Fixed.

3036662 commented:
> Also, don't you need to add `export LIBRUM_SELFHOSTED=true` to the run.sh script?

No, we don't need to export it is exported when systemd will run service ( you can look into librum-server.service file, its exported there)

3036662 commented:
> how exactly is the server supposed to find the config value

Secret values will be loaded with systemd when starting service.  You can find 

``` EnvironmentFile=/etc/librum-server/librum-server.conf ``` 

in ```librum-server.service``` file

DavidLazarescu commented:
> > Also, don't you need to add `export LIBRUM_SELFHOSTED=true` to the run.sh script?
> 
> 
> 
> No, we don't need to export it is exported when systemd will run service ( you can look into librum-server.service file, its exported there)

Ah, I see now. It probably didn't work because of the wrong file name then.

DavidLazarescu commented:
Thanks for the elaboration. I missed the line with `librum-server.conf` in the service file.

DavidLazarescu commented:
Ok, it now finds the configuration, but I am having problems with my mariadb-server installation.

How exactly would I be supposed to continue here?
![image](https://github.com/Librum-Reader/Librum-Server/assets/69865187/9ce8829e-9c03-4d7d-91c9-a79e9a0d2973)

(I have previously found a solution where I reset the password and then was able to login, but it seems to have messed up something so that the service wasnt starting anymore. I have no purged everything and reinstalled it.)

3036662 commented:
> Thanks for the elaboration. I missed the line with `librum-server.conf` in the service file.

It's possible to export all variables in .service file, but we have "secret" variables, so it's better to have them in separate file with no read-write permission for other users.

DavidLazarescu commented:
And some other small things:
- `sudo apt-get mariadb-server` doesn't work, you are missing the `install` keyword. I think we should generally use `sudo apt install ...` everywhere since this is the newer variant
- "go to /var/lib/librum-server/srv/ and add to appsettings.json and appsettings.Development.json - this block:" do we really need to add it to the `appsettings.Development.json`? We are providing instructions for a release build, so afaik. it should never be read anyways

3036662 commented:
> How exactly would I be supposed to continue here?

If your have new installation of mariadb - there is no root password, normally you simply press enter key, then you will be asked for new password for root.

DavidLazarescu commented:
And which database &Uid should I use in the connection string?
![image](https://github.com/Librum-Reader/Librum-Server/assets/69865187/097d3363-654d-4f76-8888-9c8e72488096)

3036662 commented:
> * We are providing instructions for a release build

right, it's not necessary

> * sudo apt install
ok, it will work too.



> * sudo apt-get mariadb-server

fixed

3036662 commented:
> And which database

You can write any valid db name directly to connection string, it will be created by server.

DavidLazarescu commented:
> > And which database
> 
> You can write any valid db name directly to connection string, it will be created by server.

Great, it works. I am able to call the API.

DavidLazarescu commented:
Btw, I think it might be nice to remove all of the email related stuff, such as confirming the account via a confirmation email. It is imo. annoying to need to setup SMTP credentials for this stuff.

3036662 commented:
> nice to remove

It's always easy way to delete some functionality, but not very easy to add it. 
I tried not to reduce the possibilities but to increase them.

For home use you can just push 1 value to whole database column (AspNetUsers->EmailConfirmed) all users will become verified. In my opinion - "delete unconfirmed users it's very good thing. I've created a dozen users while testing, and all of them were automatically deleted by server because of unconfirmed email.

DavidLazarescu commented:
> > nice to remove
> 
> It's always easy way to delete some functionality, but not very easy to add it. I tried not to reduce the possibilities but to increase them.
> 
> For home use you can just push 1 value to whole database column (AspNetUsers->EmailConfirmed) all users will become verified. In my opinion - "delete unconfirmed users it's very good thing. I've created a dozen users while testing, and all of them were automatically deleted by server because of unconfirmed email.

Hm, I get your point, but I don't think that its worth the extra struggle for the end user. I think by far most people who want to self-host Librum will simply setup one account and be happy with it. If they want an account deleted, they can simply delete it by logging into it and clicking "Delete Account" in the settings.

Setting up an extra SMTP service to simply use Librum is a big burden imo. so keeping it as simple as possible would be good.

3036662 commented:
> Setting up an extra SMTP service to simply use Librum is a big burden imo. so keeping it as simple as possible would be good.

I Never can argue with "Keep it simple" ! )

DavidLazarescu commented:
From what I see the server only sends emails in two places: When registering and when resetting the password.

I have already removed the email sending while registering. I'll now look into the password resetting.

DavidLazarescu commented:
The only option to avoid the email sending in the password reset is to open a browser with the form right from the server, but this is not a very nice way to handle this.
Do you have another idea?

3036662 commented:
I believe that a user who can install and configure the database and server service in linux will not be confused by the mail setup, its basics, no rocket since.
Still, the ability to reset the password and a login  tied to the mail is not bad, you can run a server, for example, for a whole house, for neighbours.
But, I won't argue.
If we continue with disabling simple checks, it will transform to simple local storage. I would consider to save full functionality.

DavidLazarescu commented:
> I believe that a user who can install and configure the database and server service in linux will not be confused by the mail setup, its basics, no rocket since. Still, the ability to reset the password and a login tied to the mail is not bad, you can run a server, for example, for a whole house, for neighbours. But, I won't argue. If we continue with disabling simple checks, it will transform to simple local storage. I would consider to save full functionality.
> 
> IMHO, the program is primarily interesting for the possibility of launching a full-fledged server.
> 
> If we disable sending email when password reset, we can same way disable password, i think. It's about local storage reader, or about server without password, i don't sure it's worth "super simple installation".

You are right, I did not think about the fact that not sending an email to the account's owner on a password reset would basically make passwords useless and thus destroy all of the security.

I will revert the last commit I made and we will keep the email sending for self-hosted servers.

DavidLazarescu commented:
Ok, it is running just fine together with the client application.
I'll do some more testing tomorrow and if everything goes well we can merge this soon.

3036662 commented:
It looks like you're putting things in order today. good job!

3036662 commented:
> Ok, everything looks good so far. Have you tested if password resetting works fine for the self-hosted server? I don't have a self-hosted email, thus I can not test it.
> 
> Can you confirm that it works fine?

You can test it with any public email service like gmail, all, etc. I' be made some test, it works fine with my email server.

3036662 commented:
> > Ok, everything looks good so far. Have you tested if password resetting works fine for the self-hosted server? I don't have a self-hosted email, thus I can not test it.
> > Can you confirm that it works fine?
> 
> You can test it with any public email service like gmail, aol, etc. I' ve made some tests, it works fine with my email server.

I mean you don't need self-hosted email. You can use any public email, if it allows smtp mail client usage.

DavidLazarescu commented:
> It looks like you're putting things in order today. good job!

Yes, pretty much everything should be done now. Could you change the PR request to merge into dev/develop instead of main though?

3036662 commented:
> > It looks like you're putting things in order today. good job!
> 
> Yes, pretty much everything should be done now. Could you change the PR request to merge into dev/develop instead of main though?

Done.
